File: rust/var-core/src/expression.rs
//! Cucumber-expression matching — the owned layer over the `cucumber-expressions`
//! crate's grammar parser. Replaces `io.cucumber.cucumberexpressions`. We take the
//! crate's AST parser (the escape-rule-dense part) and own the small, corpus-pinned
//! rest: regex generation with one named group per parameter, built-in + custom
//! parameter types, argument extraction, and `parameter_type_names`.

use crate::value::Value;
use cucumber_expressions::Expression;
use cucumber_expressions::ast::SingleExpression;
use regex::Regex;
use std::rc::Rc;

/// A parameter-type transform: maps the matched capture group(s) to a value.
pub type ParseFn = Rc<dyn Fn(&[&str]) -> Value>;

/// One captured argument of a whole-string match.
#[derive(Clone, Debug, PartialEq)]
pub struct Argument {
    /// The transformed value.
    pub value: Value,
    /// The parameter-type name (the `formats` lookup key).
    pub parameter_type_name: String,
    /// The captured group's byte offsets within the matched text (`None` if the
    /// group did not participate).
    pub group: Option<(usize, usize)>,
}

/// Registry of parameter types (built-ins + author-defined custom types).
#[derive(Clone)]
pub struct ParameterTypeRegistry {
    types: Vec<ParameterTypeDef>,
}

#[derive(Clone)]
struct ParameterTypeDef {
    name: String,
    regexp_source: String,
    transform: Transform,
}

#[derive(Clone)]
enum Transform {
    Int,
    Word,
    QuotedString,
    Custom(ParseFn),
}

// Built-in regexps, mirroring cucumber-expressions 20.0.0 (via the crate's own
// expansion). Each is wrapped in one named group per parameter at compile time.
const INT_RE: &str = r"(?:-?\d+)|(?:\d+)";
const WORD_RE: &str = r"[^\s]+";
const STRING_RE: &str = r#""[^"\\]*(?:\\.[^"\\]*)*"|'[^'\\]*(?:\\.[^'\\]*)*'"#;

impl ParameterTypeRegistry {
    /// A fresh registry with the built-in `{int}`, `{word}`, `{string}` types.
    pub fn new() -> ParameterTypeRegistry {
        ParameterTypeRegistry {
            types: vec![
                ParameterTypeDef {
                    name: "int".to_string(),
                    regexp_source: INT_RE.to_string(),
                    transform: Transform::Int,
                },
                ParameterTypeDef {
                    name: "word".to_string(),
                    regexp_source: WORD_RE.to_string(),
                    transform: Transform::Word,
                },
                ParameterTypeDef {
                    name: "string".to_string(),
                    regexp_source: STRING_RE.to_string(),
                    transform: Transform::QuotedString,
                },
            ],
        }
    }

    /// Registers a custom parameter type `name` with a bare regexp source and a
    /// transform.
    pub fn define(&mut self, name: &str, regexp_source: &str, parse: ParseFn) {
        self.types.push(ParameterTypeDef {
            name: name.to_string(),
            regexp_source: regexp_source.to_string(),
            transform: Transform::Custom(parse),
        });
    }

    fn lookup(&self, name: &str) -> Option<&ParameterTypeDef> {
        self.types.iter().find(|t| t.name == name)
    }
}

impl Default for ParameterTypeRegistry {
    fn default() -> Self {
        Self::new()
    }
}

/// An expression failed to compile.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ExpressionError {
    pub message: String,
}

/// A compiled cucumber expression.
#[derive(Clone)]
pub struct CompiledExpression {
    source: String,
    regexp_source: String,
    anchored: Regex,
    params: Vec<ParamRef>,
}

#[derive(Clone)]
struct ParamRef {
    group_name: String,
    type_name: String,
    transform: Transform,
}

impl CompiledExpression {
    /// Compiles `source` against `types`. Errors on an undefined parameter type
    /// or an un-compilable pattern.
    pub fn compile(
        source: &str,
        types: &ParameterTypeRegistry,
    ) -> Result<CompiledExpression, ExpressionError> {
        let parsed = Expression::parse(source).map_err(|e| ExpressionError {
            message: format!("failed to parse cucumber expression: {e}"),
        })?;

        let mut regex_str = String::from("^");
        let mut params = Vec::new();
        for se in &parsed.0 {
            match se {
                SingleExpression::Text(input) => regex_str.push_str(&escape_text(input.fragment())),
                SingleExpression::Whitespaces(input) => {
                    regex_str.push_str(&regex::escape(input.fragment()))
                }
                SingleExpression::Parameter(p) => {
                    let name = *p.input.fragment();
                    let def = types.lookup(name).ok_or_else(|| ExpressionError {
                        message: format!("Undefined parameter type {{{name}}}"),
                    })?;
                    let group = format!("__p{}", params.len());
                    regex_str.push_str(&format!("(?P<{group}>{})", def.regexp_source));
                    params.push(ParamRef {
                        group_name: group,
                        type_name: name.to_string(),
                        transform: def.transform.clone(),
                    });
                }
                SingleExpression::Optional(opt) => {
                    regex_str.push_str("(?:");
                    regex_str.push_str(&escape_text(opt.0.fragment()));
                    regex_str.push_str(")?");
                }
                SingleExpression::Alternation(alt) => regex_str.push_str(&alternation_regex(alt)),
            }
        }
        regex_str.push('$');

        let anchored = Regex::new(&regex_str).map_err(|e| ExpressionError {
            message: format!("failed to compile expression regex: {e}"),
        })?;
        Ok(CompiledExpression {
            source: source.to_string(),
            regexp_source: regex_str,
            anchored,
            params,
        })
    }

    /// The original expression text.
    pub fn source(&self) -> &str {
        &self.source
    }

    /// The anchored regex source (`^...$`), what the matcher strips to scan.
    pub fn regexp_source(&self) -> &str {
        &self.regexp_source
    }

    /// Matches the *entire* `text`, returning the typed arguments. `None` when
    /// `text` is not a whole match.
    pub fn match_whole(&self, text: &str) -> Option<Vec<Argument>> {
        let caps = self.anchored.captures(text)?;
        let mut args = Vec::with_capacity(self.params.len());
        for p in &self.params {
            match caps.name(&p.group_name) {
                Some(m) => args.push(Argument {
                    value: apply_transform(&p.transform, m.as_str()),
                    parameter_type_name: p.type_name.clone(),
                    group: Some((m.start(), m.end())),
                }),
                None => args.push(Argument {
                    value: Value::Null,
                    parameter_type_name: p.type_name.clone(),
                    group: None,
                }),
            }
        }
        Some(args)
    }
}

fn apply_transform(transform: &Transform, text: &str) -> Value {
    match transform {
        Transform::Int => text.parse::<i64>().map_or(Value::Null, Value::Int),
        Transform::Word => Value::String(text.to_string()),
        Transform::QuotedString => Value::String(dequote(text)),
        Transform::Custom(f) => f(&[text]),
    }
}

/// Strips a `{string}` token's surrounding quotes and unescapes `\X` → `X`.
fn dequote(s: &str) -> String {
    let chars: Vec<char> = s.chars().collect();
    if chars.len() < 2 {
        return s.to_string();
    }
    let inner = &chars[1..chars.len() - 1];
    let mut out = String::new();
    let mut i = 0;
    while i < inner.len() {
        if inner[i] == '\\' && i + 1 < inner.len() {
            out.push(inner[i + 1]);
            i += 2;
        } else {
            out.push(inner[i]);
            i += 1;
        }
    }
    out
}

/// Unescapes cucumber `\X` sequences in expression text, then regex-escapes so the
/// literal text matches verbatim.
fn escape_text(raw: &str) -> String {
    let mut unescaped = String::new();
    let mut chars = raw.chars();
    while let Some(c) = chars.next() {
        if c == '\\' {
            if let Some(n) = chars.next() {
                unescaped.push(n);
            }
        } else {
            unescaped.push(c);
        }
    }
    regex::escape(&unescaped)
}

fn alternation_regex(
    alt: &cucumber_expressions::ast::Alternation<cucumber_expressions::ast::Spanned<'_>>,
) -> String {
    use cucumber_expressions::ast::Alternative;
    let mut branches = Vec::new();
    for single in alt.0.iter() {
        let mut branch = String::new();
        for alternative in single {
            match alternative {
                Alternative::Text(t) => branch.push_str(&escape_text(t.fragment())),
                Alternative::Optional(o) => {
                    branch.push_str("(?:");
                    branch.push_str(&escape_text(o.0.fragment()));
                    branch.push_str(")?");
                }
            }
        }
        branches.push(branch);
    }
    format!("(?:{})", branches.join("|"))
}

/// Parameter-type names in source order, read from the parsed AST (escaped
/// braces `\{...\}` are literal text, not parameters).
pub fn parameter_type_names(source: &str) -> Vec<String> {
    let Ok(parsed) = Expression::parse(source) else {
        return Vec::new();
    };
    parsed
        .0
        .iter()
        .filter_map(|se| match se {
            SingleExpression::Parameter(p) => Some((*p.input.fragment()).to_string()),
            _ => None,
        })
        .collect()
}

File: go/core/expression.go
package core

import (
	"regexp"
	"strconv"
	"strings"

	ce "github.com/cucumber/cucumber-expressions-go"
)

// Cucumber-expression matching — the owned layer over the official
// cucumber-expressions-go library. We use the library to compile expressions,
// generate the anchored regexp, and extract the per-parameter capture group
// tree (values + byte offsets); we own the small, corpus-pinned rest: applying
// our own value transforms, argument extraction, and parameterTypeNames.
//
// Rationale (recorded in ADR 0010): the official Go library is stale at
// v6.2.0 rather than the 20.0.0 line every other port pins, but its built-in
// {int}/{word}/{string} regexps are byte-identical to the reference's, so it
// reproduces the goldens once we apply the reference's own transforms (a bare
// `Group.Value()` for the built-ins, the inner capture groups for a custom
// type). Only {int}, {word}, {string} and author-defined custom types are
// exercised by the corpus.

// ParseFn is a parameter-type transform. It receives the type's regexp capture
// groups in order (a group that did not participate arrives as ""); a regexp
// with no groups of its own receives the whole matched text as the single
// element. Mirrors the reference CaptureGroupTransformer.
type ParseFn func(groups []string) Value

// argument is one captured argument of a whole-string match.
type argument struct {
	value             Value
	parameterTypeName string
	// groupStart/groupEnd are the captured group's byte offsets within the
	// matched text; hasGroup is false when the group did not participate.
	hasGroup   bool
	groupStart int
	groupEnd   int
}

type transformKind int

const (
	tInt transformKind = iota
	tWord
	tQuotedString
	tCustom
)

// builtinKind returns the transform kind of a built-in parameter type, or
// (tCustom, false) if name is not a built-in.
func builtinKind(name string) (transformKind, bool) {
	switch name {
	case "int":
		return tInt, true
	case "word":
		return tWord, true
	case "string":
		return tQuotedString, true
	}
	return tCustom, false
}

// paramRef is a compiled expression's reference to one parameter's transform.
type paramRef struct {
	typeName string
	kind     transformKind
	custom   ParseFn
}

// compiledExpression is a compiled cucumber expression.
type compiledExpression struct {
	source       string
	regexpSource string // anchored ^...$
	libExpr      *ce.CucumberExpression
	params       []paramRef
}

// compileExpression compiles source against the registry's custom parameter
// types. Errors on an undefined parameter type or an un-compilable pattern.
func compileExpression(source string, customTypes []CustomParameterType, customParse map[string]ParseFn) (*compiledExpression, error) {
	libReg := ce.NewParameterTypeRegistry()
	for _, ct := range customTypes {
		re, err := regexp.Compile(ct.Regexp)
		if err != nil {
			return nil, &RegistryError{Kind: ErrExpression, Message: "failed to compile parameter type regexp: " + err.Error()}
		}
		pt, err := ce.NewParameterType(ct.Name, []*regexp.Regexp{re}, "", func(...*string) interface{} { return nil }, false, false)
		if err != nil {
			return nil, &RegistryError{Kind: ErrExpression, Message: err.Error()}
		}
		libReg.DefineParameterType(pt)
	}
	libExpr, err := ce.NewCucumberExpression(source, libReg)
	if err != nil {
		return nil, &RegistryError{Kind: ErrExpression, Message: err.Error()}
	}

	names := ParameterTypeNames(source)
	params := make([]paramRef, len(names))
	for i, name := range names {
		if kind, ok := builtinKind(name); ok {
			params[i] = paramRef{typeName: name, kind: kind}
		} else {
			params[i] = paramRef{typeName: name, kind: tCustom, custom: customParse[name]}
		}
	}

	return &compiledExpression{
		source:       source,
		regexpSource: libExpr.Regexp().String(),
		libExpr:      libExpr,
		params:       params,
	}, nil
}

// matchWhole matches the entire text, returning the typed arguments. nil when
// text is not a whole match.
func (c *compiledExpression) matchWhole(text string) []argument {
	args, err := c.libExpr.Match(text)
	if err != nil || args == nil {
		return nil
	}
	out := make([]argument, len(args))
	for i, a := range args {
		p := c.params[i]
		g := a.Group()
		if g == nil || g.Value() == nil {
			out[i] = argument{value: NullValue, parameterTypeName: p.typeName, hasGroup: false}
			continue
		}
		var val Value
		switch p.kind {
		case tInt:
			val = parseIntValue(*g.Value())
		case tWord:
			val = StrValue(*g.Value())
		case tQuotedString:
			val = StrValue(dequote(*g.Value()))
		case tCustom:
			groups := groupValues(g)
			if p.custom != nil {
				val = p.custom(groups)
			} else {
				val = NullValue
			}
		}
		out[i] = argument{
			value:             val,
			parameterTypeName: p.typeName,
			hasGroup:          true,
			groupStart:        g.Start(),
			groupEnd:          g.End(),
		}
	}
	return out
}

// groupValues returns a group's capture values (its children, or the whole
// match as a single element when there are no child groups), with a
// non-participating group rendered as "".
func groupValues(g *ce.Group) []string {
	vals := g.Values()
	out := make([]string, len(vals))
	for i, v := range vals {
		if v == nil {
			out[i] = ""
		} else {
			out[i] = *v
		}
	}
	return out
}

func parseIntValue(text string) Value {
	n, err := strconv.ParseInt(text, 10, 64)
	if err != nil {
		return NullValue
	}
	return IntValue(n)
}

// dequote strips a {string} token's surrounding quotes and unescapes \X → X.
func dequote(s string) string {
	chars := []rune(s)
	if len(chars) < 2 {
		return s
	}
	inner := chars[1 : len(chars)-1]
	var out strings.Builder
	i := 0
	for i < len(inner) {
		if inner[i] == '\\' && i+1 < len(inner) {
			out.WriteRune(inner[i+1])
			i += 2
		} else {
			out.WriteRune(inner[i])
			i++
		}
	}
	return out.String()
}

// ParameterTypeNames returns the parameter-type names in source order, reading
// the `{name}` tokens directly from source (escaped braces `\{...\}` are
// literal text, not parameters).
func ParameterTypeNames(source string) []string {
	var names []string
	runes := []rune(source)
	i := 0
	for i < len(runes) {
		c := runes[i]
		if c == '\\' {
			i += 2 // skip the escaped char
			continue
		}
		if c == '{' {
			j := i + 1
			var name strings.Builder
			for j < len(runes) && runes[j] != '}' {
				if runes[j] == '\\' && j+1 < len(runes) {
					name.WriteRune(runes[j+1])
					j += 2
					continue
				}
				name.WriteRune(runes[j])
				j++
			}
			if j < len(runes) { // found closing brace
				names = append(names, name.String())
				i = j + 1
				continue
			}
		}
		i++
	}
	return names
}

// stripAnchors strips a compiled expression's ^...$ anchors so an unanchored
// scan can find it anywhere in a sentence.
func stripAnchors(source string) string {
	s := strings.TrimPrefix(source, "^")
	s = strings.TrimSuffix(s, "$")
	return s
}
